File: python/mirage/commands/builtin/generic/du.py
from collections.abc import Awaitable, Callable

from mirage.commands.builtin.utils.formatting import _human_size
from mirage.types import PathSpec


def _format_size(size: int, human: bool) -> str:
    return _human_size(size) if human else str(size)


def _depth(entry_path: str, base_path: str) -> int:
    base = base_path.rstrip("/")
    rel = entry_path.rstrip("/")[len(base):]
    if not rel:
        return 0
    return rel.strip("/").count("/") + 1


async def _du_one(
    path: PathSpec,
    compute_total: Callable[[PathSpec], Awaitable[int]],
    compute_all: Callable[[PathSpec], Awaitable[tuple[list[tuple[str, int]],
                                                      int]]],
    *,
    s: bool,
    a: bool,
    h: bool,
    max_depth: int | None,
) -> tuple[str, int]:
    label = path.original

    if s:
        total = await compute_total(path)
        return _format_size(total, h) + "\t" + label, total

    all_entries, total = await compute_all(path)
    if not all_entries:
        total = await compute_total(path)
        return _format_size(total, h) + "\t" + label, total

    if not a:
        return _format_size(total, h) + "\t" + label, total

    entries = all_entries
    if max_depth is not None:
        entries = [(p, sz) for p, sz in entries
                   if _depth(p, label) <= max_depth]
    if not entries:
        return _format_size(total, h) + "\t" + label, total

    lines = [_format_size(sz, h) + "\t" + p for p, sz in entries]
    display_total = sum(sz for _, sz in entries)
    return "\n".join(lines), display_total


async def du(
    paths: list[PathSpec],
    *,
    compute_total: Callable[[PathSpec], Awaitable[int]],
    compute_all: Callable[[PathSpec], Awaitable[tuple[list[tuple[str, int]],
                                                      int]]],
    s: bool = False,
    a: bool = False,
    h: bool = False,
    max_depth: int | None = None,
    c: bool = False,
) -> str:
    sub_texts: list[str] = []
    totals: list[int] = []
    for p in paths:
        text, total = await _du_one(
            p,
            compute_total,
            compute_all,
            s=s,
            a=a,
            h=h,
            max_depth=max_depth,
        )
        sub_texts.append(text)
        totals.append(total)
    out = "\n".join(sub_texts)
    if c:
        grand = sum(totals)
        out += "\n" + _format_size(grand, h) + "\ttotal"
    return out


async def _du_block(
    p0: PathSpec,
    compute_total: Callable[[PathSpec], Awaitable[int]],
    compute_all: Callable[[PathSpec], Awaitable[list[tuple[str, int]]]] | None,
    *,
    s: bool,
    a: bool,
    h: bool,
    max_depth: int | None,
) -> tuple[list[str], int]:
    if s or compute_all is None:
        total = await compute_total(p0)
        return [_format_size(total, h) + "\t" + p0.original], total
    all_entries = await compute_all(p0)
    if not all_entries:
        total = await compute_total(p0)
        return [_format_size(total, h) + "\t" + p0.original], total
    if not a:
        all_entries = [(p, sz) for p, sz in all_entries if p == p0.original]
    if max_depth is not None:
        all_entries = [(p, sz) for p, sz in all_entries
                       if _depth(p, p0.original) <= max_depth]
    if not all_entries:
        total = await compute_total(p0)
        return [_format_size(total, h) + "\t" + p0.original], total
    lines = [_format_size(sz, h) + "\t" + p for p, sz in all_entries]
    return lines, sum(sz for _, sz in all_entries)


async def du_multi(
    paths: list[PathSpec],
    *,
    compute_total: Callable[[PathSpec], Awaitable[int]],
    compute_all: Callable[[PathSpec], Awaitable[list[tuple[str, int]]]]
    | None = None,
    h: bool = False,
    s: bool = False,
    a: bool = False,
    max_depth: int | None = None,
    c: bool = False,
) -> bytes:
    lines: list[str] = []
    totals: list[int] = []
    for p0 in paths:
        block, total = await _du_block(p0,
                                       compute_total,
                                       compute_all,
                                       s=s,
                                       a=a,
                                       h=h,
                                       max_depth=max_depth)
        lines.extend(block)
        totals.append(total)
    if c:
        lines.append(_format_size(sum(totals), h) + "\ttotal")
    return "\n".join(lines).encode()

File: typescript/packages/core/src/commands/builtin/github/du.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { GitHubAccessor } from '../../../accessor/github.ts'
import { du as s3Du, duAll as s3DuAll } from '../../../core/github/du.ts'
import { resolveGlob } from '../../../core/github/glob.ts'
import { IOResult, type ByteSource } from '../../../io/types.ts'
import { PathSpec, ResourceName } from '../../../types.ts'
import { command, type CommandFnResult, type CommandOpts } from '../../config.ts'
import { specOf } from '../../spec/builtins.ts'
import { humanSize } from '../utils/formatting.ts'
import { rstripSlash, stripSlash } from '../../../util/slash.ts'

const ENC = new TextEncoder()

function formatSize(size: number, human: boolean): string {
  return human ? humanSize(size) : String(size)
}

function depth(entryPath: string, basePath: string): number {
  const base = rstripSlash(basePath)
  const rel = rstripSlash(entryPath).slice(base.length)
  if (rel === '') return 0
  return (stripSlash(rel).match(/\//g) ?? []).length + 1
}

async function duCommand(
  accessor: GitHubAccessor,
  paths: PathSpec[],
  _texts: string[],
  opts: CommandOpts,
): Promise<CommandFnResult> {
  const resolved = await resolveGlob(accessor, paths, opts.index ?? undefined)
  const p0 =
    resolved[0] ??
    new PathSpec({
      original: '/',
      directory: '/',
      resolved: false,
      prefix: opts.mountPrefix ?? '',
    })
  const human = opts.flags.h === true
  const summarize = opts.flags.s === true
  const all = opts.flags.a === true
  const cumulate = opts.flags.c === true
  const maxDepthFlag =
    typeof opts.flags.max_depth === 'string'
      ? opts.flags.max_depth
      : typeof opts.flags['max-depth'] === 'string'
        ? opts.flags['max-depth']
        : null

  if (summarize) {
    const total = await s3Du(accessor, p0)
    let output = formatSize(total, human) + '\t' + p0.original
    if (cumulate) output += '\n' + formatSize(total, human) + '\ttotal'
    return [ENC.encode(output), new IOResult()]
  }
  let allEntries = await s3DuAll(accessor, p0)
  if (allEntries.length === 0) {
    const total = await s3Du(accessor, p0)
    let output = formatSize(total, human) + '\t' + p0.original
    if (cumulate) output += '\n' + formatSize(total, human) + '\ttotal'
    return [ENC.encode(output), new IOResult()]
  }
  if (!all) {
    allEntries = allEntries.filter(([p]) => p === p0.original)
  }
  if (maxDepthFlag !== null) {
    const md = Number.parseInt(maxDepthFlag, 10)
    allEntries = allEntries.filter(([p]) => depth(p, p0.original) <= md)
  }
  if (allEntries.length === 0) {
    const total = await s3Du(accessor, p0)
    let output = formatSize(total, human) + '\t' + p0.original
    if (cumulate) output += '\n' + formatSize(total, human) + '\ttotal'
    return [ENC.encode(output), new IOResult()]
  }
  const lines = allEntries.map(([p, sz]) => formatSize(sz, human) + '\t' + p)
  if (cumulate) {
    const grand = allEntries.reduce((sum, [, sz]) => sum + sz, 0)
    lines.push(formatSize(grand, human) + '\ttotal')
  }
  const out: ByteSource = ENC.encode(lines.join('\n'))
  return [out, new IOResult()]
}

export const GITHUB_DU = command({
  name: 'du',
  resource: ResourceName.GITHUB,
  spec: specOf('du'),
  fn: duCommand,
})
